File: bindings/java-jni/src/test/java/org/maplibre/nativejni/render/NativeBufferTest.java
package org.maplibre.nativejni.render;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

// Support invariant for BND-166: NativeBuffer is Java JNI's reusable CPU readback buffer,
// so its capacity and closed-state checks protect public readback ownership behavior.
class NativeBufferTest {
  @Test
  void bnd166RejectsCapacitiesBeyondDirectByteBufferLimit() {
    assertThrows(
        IllegalArgumentException.class, () -> NativeBuffer.allocate((long) Integer.MAX_VALUE + 1));
  }

  @Test
  void bnd166CloseInvalidatesWrapper() {
    var buffer = NativeBuffer.allocate(4);
    assertEquals(4, buffer.byteLength());

    buffer.close();

    assertThrows(IllegalStateException.class, buffer::byteLength);
    assertThrows(IllegalStateException.class, buffer::toByteArray);
  }
}

File: bindings/kotlin/src/jvmTest/kotlin/org/maplibre/nativeffi/render/NativeBufferTest.kt
package org.maplibre.nativeffi.render

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class NativeBufferTest {
  @Test
  fun nativeBufferTracksCapacityAndRejectsAfterClose() {
    val buffer = NativeBuffer.allocate(4L)

    assertEquals(4L, buffer.byteLength())
    assertEquals(4, buffer.toByteArray().size)
    buffer.ensureCapacity(4L)
    assertFailsWith<IllegalArgumentException> { buffer.ensureCapacity(5L) }

    buffer.close()
    buffer.close()
    assertFailsWith<IllegalStateException> { buffer.byteLength() }
  }

  @Test
  fun closeDuringBorrowRejectsNewAccessAndCompletesAfterBorrowReturns() {
    val buffer = NativeBuffer.allocate(4L)

    buffer.borrow { _, length ->
      assertEquals(4L, length)
      buffer.close()
      assertFailsWith<IllegalStateException> { buffer.byteLength() }
    }

    buffer.close()
    assertFailsWith<IllegalStateException> { buffer.toByteArray() }
  }

  @Test
  fun zeroLengthBufferHasNoBytes() {
    NativeBuffer.allocate(0L).use { buffer ->
      assertEquals(0L, buffer.byteLength())
      assertEquals(0, buffer.toByteArray().size)
    }
  }
}
